snippets/src/language/snippet_context.rs

```diff
@@ -0,0 +1,56 @@
+use crate::impl_prelude::SnippetLocation;
+use getset::{CopyGetters, Getters};
+use tree_sitter::Node;
+
+/// This structure represents a view into a larger piece of parsed text.
+/// For snippet scanning, we generally look at just parts of a larger piece of text for each snippet.
+/// However, the parsed nodes all reference locations in the original text.
+/// This structure is meant to make it easier to find content inside a snippet based on previously extracted nodes.
+#[derive(Debug, PartialEq, Getters, CopyGetters)]
+pub struct SnippetContext<'a> {
+    offset: usize,
+    /// The location in the original text of this snippet.
+    #[getset(get = "pub")]
+    location: SnippetLocation,
+    /// The nodes that have been parsed from this context.
+    #[getset(get = "pub")]
+    context_nodes: Vec<Node<'a>>,
+    /// The slice of text represented by [`SnippetLocation`].
+    #[getset(get_copy = "pub")]
+    content: &'a [u8],
+}
+
+impl<'a> SnippetContext<'a> {
+    /// Make a new SnippetContext from a sequence of nodes, a location within the original parsed text, and a sequence of bytes for data inside the snippet.
+    /// There is no checking or guarantee that the provided nodes fall within the bounds of the provided content.
+    pub fn new(context_nodes: Vec<Node<'a>>, location: SnippetLocation, content: &'a [u8]) -> Self {
+        let crate::ByteOffset(offset) = location.byte_offset;
+        SnippetContext {
+            offset,
+            context_nodes,
+            location,
+            content,
+        }
+    }
+
+    /// Get content from the snippet which is not in ranges covered by the provided nodes.
+    pub fn retrieve_negative_content(
```

Comment:
I'm not 100% sure about this name

What this function is doing is removing the provided nodes

Maybe `without_provided_nodes` or `remove_nodes`?

Reply:
I think those names might imply that we are altering the `SnippetContext` itself which is not correct. I think your point around the name being confusing is a good one. I'll have to think about it.

I think that it's generally good to have a structure that represents this concept, but I am still a little uncomfortable with the interface I've defined for it. I think ideally, this function would return a whole new `SnippetContext` with a new view into the modified context. Then the interface is effectively `context.filter_nodes(|f| <some predicate>)`. 

Initially I was thinking it'd be hard to keep track of all the offsets as nodes are added and removed. But actually if someone ever tries to view the content I can just generate it on-demand from the original. Maybe this is more possible than I thought it was.

Reply:
I spent a little time with that above and found that it doesn't work so well. `Node`s can overlap, so filtering nodes and outputting content from the remaining ones doesn't do what you'd expect. For example:
```
int main () {
// I am a comment
}
```
Has a node of kind comment which contains the comment text. But it also has a node called "translation_unit" that covers the entire parsed function, comment and all. So simply removing the comments and copying the remaining text out doesn't work in a straightforward way. I think this is probably why I did such a weird solution to begin with.

For now, I've just changed the name to `retrieve_content_around_nodes`.